nevergrad/optimization/optimizerlib.py

```diff
@@ -1203,68 +1203,74 @@ def _internal_tell_candidate(self, candidate: base.Candidate, value: float) -> N
+            if self.num_ask < sum_budget:
+                break
+        # if we are over budget, then use the last one...
+        return opt.ask()
+
+    def _internal_tell_candidate(self, candidate: base.Candidate, value: float) -> None:
+        # Let us inform all concerned algorithms
+        sum_budget = 0.0
+        for opt in self._optimizers:
+            sum_budget += float("inf") if opt.budget is None else opt.budget
+            if self.num_tell < sum_budget:
+                opt.tell(candidate, value)
+
+
+class Chaining(base.ParametrizedFamily):
     """
     A chaining consists in running algorithm 1 during T1, then algorithm 2 during T2, then algorithm 3 during T3, etc.
-    Each algorithm (except the first one) is fed with the entire archive.
-    """
-    
-    def __init__(
-        self,
-        list_of_algorithms,
-        list_of_budgets,
-    ) -> None:
-        # Either we have the budget for each algorithm, or the last algorithm uses the rest of the budget, so:
-        assert len(list_of_algorithms) == len(list_of_budgets) or len(list_of_algorithms) == len(list_of_budgets) + 1 
+    Each algorithm is fed with what happened before it.
 
-        # Let us store the two key parameters:
-        self._list_of_budgets = list_of_budgets
-        self._list_of_algorithms = list_of_algorithms
+    Parameters
+    ----------
+    optimizers: list of Optimizer classes
+        the sequence of optimizers to use
+    budgets: list of int
+        the corresponding budgets for each optimizer but the last one
 
+    """
+    _optimizer_class = _Chain
+
+    def __init__(self, optimizers: Sequence[Union[base.OptimizerFamily, Type[base.Optimizer]]],
+                 budgets: Sequence[Union[str, int]]) -> None:
+        # Either we have the budget for each algorithm, or the last algorithm uses the rest of the budget, so:
+        self.budgets = tuple(budgets)
+        self.optimizers = tuple(optimizers)
+        assert len(self.optimizers) == len(self.budgets) + 1
+        assert all(x in ("dimension", "num_workers") or x > 0 for x in self.budgets)  # type: ignore
+        super().__init__()
 
-    # From the user point of view, the call below is the creation of the optimizer.
-    def __call__(self, instrumentation: Union[int, Instrumentation], budget: Optional[int] = None, num_workers: int = 1) -> base.Optimizer:
 
-        list_of_algorithms = self._list_of_algorithms
-        list_of_budgets = self._list_of_budgets
-
-        # We create the class we want to use.
-        class chain(base.Optimizer):
-
-            def __init__(self, instrumentation: Union[int, Instrumentation], budget: Optional[int] = None, num_workers: int = 1) -> None:
-                super().__init__(instrumentation, budget=budget, num_workers=num_workers)
-                self._list_of_budgets = list_of_budgets
-                self._list_of_algorithms = [l(instrumentation, budget, num_workers) for l in list_of_algorithms]
```

Comment:
each optimizer should be initialized with its own budget: initializing LHS with budget 100 if it will be asked only 5 times is suboptimal, since only 5 of the points will be asked and there would not be any warranty on their distribution

Reply:
Argh good catch